components/ai_chat/renderer/page_content_extractor.cc

```diff
@@ -0,0 +1,223 @@
+        if (nodes.length) {
+          let selectedNode = nodes[0]
+          for (const node of nodes) {
+            if (node.srclang.toLowerCase() === 'en') {
+              selectedNode = node
+            }
+          }
+          return node.src
+        }
+      })()
+    )JS";
+
+constexpr auto kYouTubeHosts = base::MakeFixedFlatSet<base::StringPiece>(
+    {"youtube.com", "www.youtube.com", "m.youtube.com"});
+
+// TODO(petemill): Use heuristics to determine if page's main focus is
+// a video, and not a hard-coded list of Url hosts.
+constexpr auto kVideoTrackHosts =
+    base::MakeFixedFlatSet<base::StringPiece>({"www.ted.com"});
+
+}  // namespace
+
+namespace ai_chat {
+
+PageContentExtractor::PageContentExtractor(
+    content::RenderFrame* render_frame,
+    service_manager::BinderRegistry* registry,
+    int32_t global_world_id,
+    int32_t isolated_world_id)
+    : content::RenderFrameObserver(render_frame),
+      global_world_id_(global_world_id),
+      isolated_world_id_(isolated_world_id),
+      weak_ptr_factory_(this) {
+  if (!render_frame->IsMainFrame()) {
+    return;
+  }
+  // Bind mojom API to allow browser to communicate with this class
+  // Being a RenderFrameObserver, this object is scoped to the RenderFrame.
+  // Unretained is safe here because `registry` is also scoped to the
+  // RenderFrame.
+  registry->AddInterface(base::BindRepeating(
+      &PageContentExtractor::BindReceiver, base::Unretained(this)));
+}
+
+PageContentExtractor::~PageContentExtractor() = default;
+
+void PageContentExtractor::OnDestruct() {
+  delete this;
+}
+
+void PageContentExtractor::ExtractPageContent(
+    mojom::PageContentExtractor::ExtractPageContentCallback callback) {
+  VLOG(1) << "AI Chat renderer has been asked for page content.";
+  blink::WebLocalFrame* main_frame = render_frame()->GetWebFrame();
+  GURL origin =
+      url::Origin(((const blink::WebFrame*)main_frame)->GetSecurityOrigin())
+          .GetURL();
+
+  // Decide which technique to use to extract content from page...
+  // 1) Video - YouTube's custom link to transcript
+  // 2) Video - <track> element specifying text location
+  // 3) Text - find the "main" text of the page
+
+  if (origin.is_valid()) {
+    std::string host = origin.host();
+    if (base::Contains(kYouTubeHosts, host)) {
+      VLOG(1) << "YouTube transcript type";
+      // Do Youtube extraction
+      v8::HandleScope handle_scope(v8::Isolate::GetCurrent());
+      blink::WebScriptSource source = blink::WebScriptSource(
+          blink::WebString::FromUTF16(kYoutubeTranscriptUrlExtractionScript));
+      auto script_callback = base::BindOnce(
+          &PageContentExtractor::OnJSTranscriptUrlResult,
+          weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+          ai_chat::mojom::PageContentType::VideoTranscriptYouTube);
+      // Main world so that we can access a global variable
+      main_frame->RequestExecuteScript(
```

Comment:
There are differences other than script source that different kinds of transcript extraction strategies may have. We have 2 strategies right now, but I imagine we'll have a few. Differences when using script execution could be:
- Different worlds (these 2 scrips do have different worlds already)
- Wait for a promise or wait for a callback
In fact, some strategies might not run RequestExecuteScript at all and might simply use C++ to do some DOM lookups. I think that's the main point to keep them separate as it seems like the order of deciding which strategy to use is more important than combining code simply because it's similar. 
I did leave comments before each to note the differences (although looks like I put "main world" for both in the comments, instead of "isolated world" for the non-youtube one), and I can make the comments more obvious.